packages/wallet-sdk/src/connection/WalletLinkConnection.ts

```diff
@@ -101,111 +73,98 @@ export class WalletLinkConnection {
     private diagnostic?: DiagnosticLogger,
     WebSocketClass: typeof WebSocket = WebSocket
   ) {
-    const ws = new RxWebSocket<ServerMessage>(`${linkAPIUrl}/rpc`, WebSocketClass);
+    const ws = new WalletLinkWebSocket(`${linkAPIUrl}/rpc`, WebSocketClass);
     this.ws = ws;
 
-    // attempt to reconnect every 5 seconds when disconnected
-    this.subscriptions.add(
-      ws.connectionState$
-        .pipe(
-          tap(
-            (state) =>
-              this.diagnostic?.log(EVENTS.CONNECTED_STATE_CHANGE, {
-                state,
-                sessionIdHash: Session.hash(sessionId),
-              })
-          ),
-          // ignore initial DISCONNECTED state
-          skip(1),
+    this.ws.setConnectionStateListener(async (state) => {
+      // attempt to reconnect every 5 seconds when disconnected
+      this.diagnostic?.log(EVENTS.CONNECTED_STATE_CHANGE, {
+        state,
+        sessionIdHash: Session.hash(sessionId),
+      });
+
+      let connected = false;
+      switch (state) {
+        case ConnectionState.DISCONNECTED:
           // if DISCONNECTED and not destroyed
```

Comment:
Checking if it's intentional we don't skip the initial disconnect event.

Reply:
for the record:
with rxjs subject, it used to emit a default value `DISCONNECTED` whenever the observable was subscribed
by replacing it with callback listener, we expect `connectionStateListener` only gets invoked when there’s actual state changes. so we don’t need `skip(1)` for `DISCONNECTED`